llumnix/backends/vllm/migrate_backend.py

```diff
@@ -0,0 +1,257 @@
+                for idx, block_num in enumerate(blocks):
+                    self.gpu_cache[layer_idx][0][block_num].copy_(dummy_key[idx][layer_idx], non_blocking=True)
+                    self.gpu_cache[layer_idx][1][block_num].copy_(dummy_value[idx][layer_idx], non_blocking=True)
+        torch.cuda.Stream.synchronize(self.migration_stream)
+
+class RayMigrateBackend(MigrateBackendBase):
+    def __init__(self, migrate_config: MigrationConfig, cache_engine: CacheEngine, ray_world_size, ray_rank, \
+                  local_rank, scheduling_strategy, dtype, is_driver_worker, gpu_cache) -> None:
+        super().__init__()
+
+        self.migrate_config = migrate_config
+        self.cache_engine = cache_engine
+        self.num_migration_cache_blocks = migrate_config.migration_cache_blocks
+
+        self.backend = migrate_config.migration_backend
+        self.ray_world_size = ray_world_size
+        self.ray_rank = ray_rank
+        self.group_name = None
+
+        self.local_rank = local_rank
+        self.actor = ProxyActor.options(scheduling_strategy=scheduling_strategy).remote()
+        self.dtype = dtype
+        self.is_driver_worker = is_driver_worker
+        self.gpu_cache = gpu_cache
+
+        self.migration_cache_size = self.cache_engine.block_size * self.cache_engine.num_heads * self.cache_engine.head_size
+
+        if self.backend == 'gloo':
+            self.cache_device = "cpu"
+        elif self.backend == 'nccl':
+            self.cache_device = torch.device(f"cuda:{self.local_rank}")
+        else:
+            raise ValueError("backend must be 'gloo' or 'nccl'")
+
+        pin_memory = self.backend == 'gloo'
+        self.dummy_cache = torch.empty(
+            size=(2*self.num_migration_cache_blocks, self.cache_engine.num_layers, self.migration_cache_size),
+            dtype=self.cache_engine.dtype,
+            device=self.cache_device,
+            pin_memory=pin_memory
+        )
+
+        self.migration_stream = torch.cuda.Stream()
+
+    def init_col(self, name, world_size, rank) -> None:
+        self.group_name = name
+        self.ray_world_size = world_size
+        self.ray_rank = rank
+
+        col.init_collective_group(world_size=self.ray_world_size, rank=self.ray_rank,
+            backend=self.backend, group_name=self.group_name)
+
+        logger.info("create ray collective group success (group_name:{}, world_size: {}, rank: {}, backbend: {})."
+                    .format(self.group_name, self.ray_world_size, self.ray_rank, self.backend))
+
+    def warmup(self) -> None:
+        if self.ray_world_size > 1:
+            col.allreduce(self.dummy_cache[0], self.group_name)
+
+        logger.info("ray collective group warmup success (group_name:{}, world_size: {}, rank: {}, backbend: {})."
+                    .format(self.group_name, self.ray_world_size, self.ray_rank, self.backend))
+
+    def destory_col(self) -> None:
+        if self.group_name is not None:
```

Comment:
Why not assert group_name is not None here?

Reply:
Is it possible that destroy_col is called simultanously due to multi-threading?

Reply:
It's not multi-threading problem. Assert generates an error. Running destory_col when init is not executed does not affect the system's normal operation, so it is not chosen to generate an error here.